h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -0,0 +1,251 @@
+      return buildGamFrame(numGamFrame, _gamFrameKeys, _train, _parms._response_column); // add gam cols to _train
+    }
+
+    @Override
+    public void computeImpl() {
+      init(true);     //this can change the seed if it was set to -1
+      if (error_count() > 0)   // if something goes wrong, let's throw a fit
+        throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(GAM.this);
+
+      _job.update(0, "Initializing model training");
+
+      buildModel(); // build gam model 
+    }
+
+    public final void buildModel() {
+      GAMModel model = null;
+      DataInfo dinfo = null;
+      Frame newTFrame=null, gamColsFrame=null, gamColsFrameCenter=null;
+      try {
+        _job.update(0, "Adding GAM columns to training dataset...");
+        int trainCols = _train.numCols();  // number of columns before adding GAM columns
+        String[] trainColNames = _train.names(); // column names before adding GAM columns
+        Frame newTrain = rebalance(adaptTrain(), false, _result+".temporary.train"); // add and store gam cols without centering
+        dinfo = new DataInfo(newTrain.clone(), _valid, 1, _parms._use_all_factor_levels || _parms._lambda_search, DataInfo.TransformType.NONE, DataInfo.TransformType.NONE,
+                _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.Skip,
+                _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.MeanImputation || _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.PlugValues,
+                _parms.makeImputer(),
+                false, hasWeightCol(), hasOffsetCol(), hasFoldCol(), null);
+        DKV.put(dinfo._key, dinfo);
+        model = new GAMModel(dest(), _parms, new GAMModel.GAMModelOutput(GAM.this, dinfo._adaptedFrame, dinfo));
+        model.delete_and_lock(_job);
+        newTFrame = _centerGAM?new Frame(buildGamFrameCenter(_parms._gam_X.length, _gamFrameKeysCenter, _parms.train(), 
+                _parms._response_column, _parms.train().numCols(), _parms.train().names())):new Frame(dinfo._adaptedFrame);  // get frames with correct predictors and spline functions
+        DKV.put(newTFrame); // This one will cause deleted vectors if add to Scope.track
+        if (_parms._saveGamCols) {  // save gam column keys
+          gamColsFrame = saveGAMFrames(newTrain);
+          DKV.put(gamColsFrame);  // do not Scope.track this, will cause null frame
+          model._output._gamTransformedTrain = gamColsFrame._key;
+          if (_centerGAM) {
+            gamColsFrameCenter = saveGAMFrames(newTFrame);
+            DKV.put(gamColsFrameCenter);
+            model._output._gamTransformedTrainCenter = gamColsFrameCenter._key;
+            model._output._gamGamXCenter = newTFrame._key;
+          }
+        }
+        
+        _job.update(1, "calling GLM to build GAM model...");
+        GLMModel glmModel = buildGLMModel(_parms, newTFrame); // obtained GLM model
```

Comment:
call glm to solve for coefficients with predictors and basis functions with zero-mean constraints applied to.